### PR TITLE
fix: new implementation of useSyncExternalStore

### PR DESCRIPTION
**Pull request by aidenybai**
**Please describe the changes this PR makes and why it should be merged:**

**Status**

- [x] Code changes have been tested against prettier, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the codebase
  - x ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
  - [ ] This PR changes the internal workings with no modifications to the external API (bug fixes, performance improvements)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: packages/react/hooks.ts [hunk 1]
```diff
@@ -232,11 +232,14 @@ export const useDeferredValue = (value: any) => {
...
-  const state = useState(getSnapshot());
+  const [, updateState] = useState();
+  const forceUpdate = useCallback(() => updateState({}), []);
+  const [subscriber, setSubscriber] = useState(() => subscribe(forceUpdate));
+
   useEffect(() => {
-    subscribe(state);
-  });
-  subscribe(state);
+    if (subscriber) subscriber();
```
Comment (QuiiBz) [inline packages/react/hooks.ts:240]: Can be simplified and maybe save some bytes?
```suggestion
    subscriber?.();
```

**Review by QuiiBz — APPROVED** [review]
One small suggestion, otherwise LGTM!

**Review by SukkaW — CHANGES_REQUESTED** [review]
The implementation still seems wrong.

IIRC in react's implementation, `useSyncExternalStore` will unsubscribe and then re-subscribe if `subscribe` function changes, and bail out re-subscribe if `subscribe` doesn't change.

So it is not the `getSnapshot()` but the `subscribe` that should be in the `useEffect`'s dependency array.

Also, some other existing implementation that we could learn from:

- preact's implementation: https://github.com/preactjs/preact/blob/6b92b1fab41599e6da4f96d65b07fdbe0b6ff2fc/compat/src/index.js#L136-L154
- react's `use-sync-external-store` polyfill implementation: https://github.com/facebook/react/blob/main/packages/use-sync-external-store/src/useSyncExternalStoreShimClient.js

File: packages/react/hooks.ts [hunk 2]
```diff
@@ -232,11 +232,14 @@ export const useDeferredValue = (value: any) => {
...
+  const [subscriber, setSubscriber] = useState(() => subscribe(forceUpdate));
+
   useEffect(() => {
-    subscribe(state);
-  });
-  subscribe(state);
+    subscriber?.();
+    setSubscriber(subscribe(forceUpdate));
+  }, [getSnapshot()]);
   return state;
```
Comment (SukkaW) [inline packages/react/hooks.ts:243]: Where does the `state` come from if `const state = useState(getSnapshot());` is removed?

Comment (aidenybai) [reply packages/react/hooks.ts:263]: I think we can go ahead with copying Preact's implementation.

**Comment by aidenybai** [comment]
I see, thank you for the helpful explanation @SukkaW. I think we should go ahead and copy Preact in this case (thank you preactjs!) to be safe.

**Comment by QuiiBz** [comment]
> ...copy Preact in this case...

Let's not forget to credit them inside the file :)

**Comment by aidenybai** [comment]
good catch, adding attribution

**Review by SukkaW — APPROVED** [review]

**Pull request merged**